src/server/util/api.js

```diff
@@ -12,11 +12,45 @@ exports.apiError = (res, error) => {
     return res.status(status).json(result);
 };
 
+// error handling middleware needs 4 parameters, even though we don't use it, we MUST define it:
+// eslint-disable-next-line no-unused-vars
+exports.errorHandler = (error, req, res, next) => exports.apiError(res, error);
+
 // This function wraps around async middlewares and catches errors
-exports.asyncMiddleware = middle => async (req, res, next) => {
-    try {
-        return await middle(req, res, next);
-    } catch (error) {
-        exports.apiError(res, error);
+exports.asyncMiddleware = middle => {
+    if (typeof middle !== 'function') {
+        throw new Error('Must provide a function to asyncMiddleware');
+    }
+    if (middle.length == 2 || middle.length == 3) {
+        // (req, res) or (req, res, next) -- express handles both the same, assuming
+        // middle won't call next() if it's techincally the "endpoint" so the function we return
+        // can have 3 arguments or 2 arguments and be treated the same
+        return async function asyncMiddlewareWrapper(req, res, next) {
```

Comment:
I'm a little confused picturing how this runs. Does this mean the middleware is creating a new function every time it runs? Does it run each time its route is hit? Will this extra work get expensive?

Reply:
Nope, this is standard "function returning a function" stuff -- the outer scope is maintained, and this "closure" here is the middleware passed into express, we only create the function ONCE when we call `asyncMiddleware(async f(req, res) {})` and it returns `asyncMiddlewareWrapper(req, res, next)`  which holds the outer reference to the passed in `f` as `middle`

Reply:
OOhhh okay cool, that makes sense 👍